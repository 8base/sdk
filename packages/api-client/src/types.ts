import { TableSchema } from '@8base/utils';
import * as filestack from 'filestack-js';

export type SchemaResponse = {
  tablesList: {
    items: TableSchema[];
  };
};

export type FileUploadInfoResponse = {
  policy: string;
  signature: string;
  apiKey: string;
  path: string;
};

export type FilestackClient = ReturnType<typeof filestack.init>;
