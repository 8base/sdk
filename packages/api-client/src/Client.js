// @flow
import type { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { GraphQLClient } from 'graphql-request';
import { hasIdTokenExpiredError } from '@8base/utils';
import * as R from 'ramda';

const USER_REFRESH_TOKEN_QUERY = gql`
  mutation UserRefreshToken($refreshToken: String!, $email: String!) {
    userRefreshToken(data: {
      refreshToken: $refreshToken,
      email: $email,
    }) {
      refreshToken
      idToken
    }
  }
`;

const getRefreshToken = R.path(['userRefreshToken', 'refreshToken']);
const getIdToken = R.path(['userRefreshToken', 'idToken']);

type GraphQLClientResponse = {
  errors: Array<Object>,
  data: Object,
};

type GraphQLClientRequest = {
  query: DocumentNode,
  variables: Object,
};

type GraphQLClientError = {
  response: GraphQLClientResponse,
  request: GraphQLClientRequest,
};

class RefreshTokenInvalidError extends Error {
  constructor() {
    super('Can\'t refresh token.');
  }
}

/**
 * Client provides methods to make requests to the API.
 * @param endpoint - The endpoint which are using for make requests.
 */
class Client {
  gqlc: GraphQLClient;

  workspaceId: string;
  email: string;
  idToken: ?string;
  refreshToken: string;

  constructor(endpoint: string) {
    this.gqlc = new GraphQLClient(endpoint);
  }

  /**
   * Update id token.
   * @param idToken - The id token.
   */
  setIdToken(idToken: ?string) {
    this.idToken = idToken;

    if (idToken) {
      this.gqlc.setHeader('Authorization', `Bearer ${idToken}`);
    } else {
      delete this.gqlc.options.headers.Authorization;
    }
  }

  /**
   * Update refresh token.
   * @param refreshToken - The refresh token.
   */
  setRefreshToken(refreshToken: string) {
    this.refreshToken = refreshToken;
  }

  /**
   * Update user email.
   * @param email - The user email.
   */
  setEmail(email: string) {
    this.email = email;
  }

  /**
   * Update workspace identifier.
   * @param workspaceId - The workspace identifier.
   */
  setWorkspaceId(workspaceId: string) {
    this.workspaceId = workspaceId;
    this.gqlc.setHeader('workspace', workspaceId);
  }

  async tryToRefreshToken(err: GraphQLClientError) {
    const { refreshToken, email } = this;

    this.setIdToken(null);

    let response = null;

    try {
      response = await this.gqlc.request(USER_REFRESH_TOKEN_QUERY, {
        refreshToken,
        email,
      });
    } catch (err) {
      throw new RefreshTokenInvalidError();
    }

    const newRefreshToken = getRefreshToken(response);
    const newIdToken = getIdToken(response);

    this.setRefreshToken(newRefreshToken);
    this.setIdToken(newIdToken);

    return this.request(err.request.query, err.request.variables);
  }

  handleRequestErrors = (err: GraphQLClientError) => {
    // $FlowFixMe
    if (hasIdTokenExpiredError(R.pathOr([], ['response', 'errors'], err))) {
      return this.tryToRefreshToken(err);
    }
    throw err;

  };

  /**
   * Send request to the API
   * @param query - GraphQL query.
   * @param variables - The variables that will be used when executing the query.
   * @returns {Promise}
   */
  request(query: string | DocumentNode, variables: Object = {}) {
    return this.gqlc.request(query, variables).catch(this.handleRequestErrors);
  }
}

export { Client };
