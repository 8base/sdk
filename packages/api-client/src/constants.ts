export const USER_REFRESH_TOKEN_QUERY = `
  mutation UserRefreshToken($refreshToken: String!, $email: String!) {
    userRefreshToken(data: {
      refreshToken: $refreshToken,
      email: $email,
    }) {
      refreshToken
      idToken
    }
  }
`;

export const TABLES_LIST_QUERY = `
  query TableList($filter: TableListFilter) {
    tablesList(filter: $filter) {
      items {
        ...TableFragment
      }
    }
  }

  fragment TableFragment on Table {
    id
    name
    displayName
    isSystem
    fields {
      ...TableFieldFragment
    }
  }

  fragment TableFieldFragment on TableField {
    ...CommonTableFieldFragment
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
      ...SmartFieldTypesAttributes
    }
  }

  fragment CommonTableFieldFragment on TableField {
    id
    name
    displayName
    description
    fieldType
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
    }
    isList
    isRequired
    isUnique
    defaultValue
    isSystem
    isMeta
    relation {
      refFieldName
      refFieldDisplayName
      relationTableName
      relationFieldName
      refTable {
        id
        name
      }
      refFieldIsList
      refFieldIsRequired
    }
  }

  fragment DateFieldTypeAttributes on DateFieldTypeAttributes {
    format
  }

  fragment TextFieldTypeAttributes on TextFieldTypeAttributes {
    format
    fieldSize
  }

  fragment NumberFieldTypeAttributes on NumberFieldTypeAttributes {
    format
    precision
    currency
    minValue
    maxValue
    isBigInt
  }

  fragment FileFieldTypeAttributes on FileFieldTypeAttributes {
    format
    maxSize
    typeRestrictions
  }

  fragment SmartFieldTypesAttributes on SmartFieldTypeAttributes {
    format
    innerFields {
      name
      displayName
      description
      fieldType
      isList
      isRequired
      isUnique
      fieldTypeAttributes {
        ...TextFieldTypeAttributes
        ...NumberFieldTypeAttributes
        ...FileFieldTypeAttributes
        ...DateFieldTypeAttributes
        ...SwitchFieldTypeAttributes
      }
    }
  }

  fragment SwitchFieldTypeAttributes on SwitchFieldTypeAttributes {
    format
    listOptions
  }
`;

export const TABLE_CREATE_MUTATION = `
  mutation TableCreate($data: TableCreateInput!) {
    tableCreate(data: $data) {
      id
    }
  }
`;

export const FIELD_CREATE_MUTATION = `
  mutation FieldCreate($data: TableFieldCreateInput!) {
    fieldCreate(data: $data) {
      id
    }
  }
`;

export const USER_QUERY = `
  query User {
    user {
      id
      email
    }
  }
`;

export const FILE_UPLOAD_INFO_QUERY = `
  query FileUploadInfo {
    fileUploadInfo {
      apiKey
      policy
      signature
      path
    }
  }
`;
