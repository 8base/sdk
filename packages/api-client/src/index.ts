export { importTables } from './importTables';
export { importData } from './importData';
export { importWorkspace } from './importWorkspace';
export { exportTables } from './exportTables';
export { exportData } from './exportData';
export { Client } from './Client';
