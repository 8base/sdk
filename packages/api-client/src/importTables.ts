import * as R from 'ramda';
import errorCodes from '@8base/error-codes';
import { Schema, FieldSchema } from '@8base/utils';
import { DocumentNode } from 'graphql';

import { TABLES_LIST_QUERY, TABLE_CREATE_MUTATION, FIELD_CREATE_MUTATION } from './constants';
import { SchemaResponse } from './types';

type ImportTablesOptions = {
  debug?: boolean;
};

const handleError = (message: string, e: { response: any }, debug: boolean) => {
  // tslint:disable-next-line no-console
  console.warn(message);

  if (debug) {
    // tslint:disable-next-line no-console
    console.warn(JSON.stringify(e.response, null, 2));
  } else if (R.pathEq(['response', 'errors', 0, 'code'], errorCodes.ValidationErrorCode, e)) {
    // tslint:disable-next-line no-console
    console.warn(JSON.stringify(R.path(['response', 'errors', 0, 'details'], e), null, 2));
  } else {
    // tslint:disable-next-line no-console
    console.warn(R.path(['response', 'errors', 0, 'message'], e));
  }
};

export const importTables = async (
  request: <T extends object>(query: string | DocumentNode, variables?: object) => Promise<T>,
  schema: Schema,
  options: ImportTablesOptions = {},
) => {
  const debug: boolean = R.propOr(false, 'debug', options);

  for (const table of schema) {
    try {
      await request(TABLE_CREATE_MUTATION, {
        data: {
          displayName: table.displayName,
          name: table.name,
        },
      });
    } catch (e) {
      handleError(`\nCan't create table "${table.name}"`, e, debug);
    }
  }

  const tablesListData = await request<SchemaResponse>(TABLES_LIST_QUERY, {
    filter: {
      onlyUserTables: false,
    },
  });

  const tablesByName = R.indexBy(R.prop('name'), tablesListData.tablesList.items);

  for (const schemaTable of schema) {
    if (schemaTable.isSystem) {
      continue;
    }

    const table = tablesByName[schemaTable.name];

    for (const schemaField of schemaTable.fields) {
      if (schemaField.isSystem) {
        continue;
      }

      const field: Partial<FieldSchema> = R.pick(
        [
          'name',
          'displayName',
          'fieldType',
          'isList',
          'isRequired',
          'isUnique',
          'defaultValue',
          'description',
          'fieldTypeAttributes',
          'relation',
          'tableId',
        ],
        schemaField,
      );

      if (field.fieldTypeAttributes) {
        field.fieldTypeAttributes = R.pick(
          [
            'format',
            'fieldSize',
            'listOptions',
            'precision',
            'currency',
            'minValue',
            'maxValue',
            'maxSize',
            'typeRestrictions',
            'isBigInt',
          ],
          field.fieldTypeAttributes,
        );
      }

      if (field.relation && field.relation.refTable.name) {
        const refTableName = field.relation.refTable.name;

        const refTable = tablesByName[refTableName];

        // @ts-ignore. Check what does do relations prop.
        if (refTable.relations && refTable.relations[schemaTable.name]) {
          // @ts-ignore. Check what does do relations prop.
          if (refTable.relations[schemaTable.name].includes(field.name)) {
            continue;
          }
        }

        field.relation = {
          ...R.pick(['refFieldIsList', 'refFieldIsRequired', 'refFieldName', 'refFieldDisplayName'], field.relation),
          refTableId: refTable.id,
        } as any;

        // @ts-ignore. Check what does relations prop do.
        table.relations = table.relations || {};

        // @ts-ignore. Check what does relations prop do.
        if (table.relations[refTableName]) {
          // @ts-ignore. Check what does relations prop do.
          table.relations[refTableName].push(field.relation.refFieldName);
        } else {
          // @ts-ignore. Check what does relations prop do.
          table.relations[refTableName] = [field.relation.refFieldName];
        }
      }

      // @ts-ignore. Check what does tableId prop do.
      field.tableId = table.id;

      try {
        await request(FIELD_CREATE_MUTATION, {
          data: field,
        });
      } catch (e) {
        handleError(`\nCan't create field "${field.name}"`, e, debug);
      }
    }
  }
};
