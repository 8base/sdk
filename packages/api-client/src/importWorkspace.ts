import { DocumentNode } from 'graphql';
import * as R from 'ramda';

import { importData } from './importData';
import { importTables } from './importTables';

export const importWorkspace = async (
  request: <T extends any>(query: string | DocumentNode, variables?: object) => Promise<T>,
  workspace: object,
) => {
  await importTables(request, R.propOr({}, 'tables', workspace));
  await importData(request, R.propOr({}, 'data', workspace));
};
