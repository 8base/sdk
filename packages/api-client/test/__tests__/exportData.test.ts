import * as nock from 'nock';

import { Client, exportData } from '../../src';
import { TABLES, FILES_TABLE, USER_TABLES, CLIENTS_TABLE, SYSTEM_DATA, DATA } from '../__fixtures__';

beforeEach(() => {
  nock.cleanAll();
});

const tablesResponse = () => ({
  data: {
    tablesList: {
      items: [FILES_TABLE],
    },
  },
});

describe('exportData', () => {
  const client = new Client('https://api.test.8base.com');

  client.setIdToken('idToken');
  client.setWorkspaceId('workspaceId');

  it('should export data from workspace with system tables', async () => {
    const mocks = [
      global.mockRequest('https://api.test.8base.com', 200, {
        data: {
          usersList: {
            items: SYSTEM_DATA.users,
          },
        },
      }),
      global.mockRequest('https://api.test.8base.com', 200, tablesResponse),
      global.mockRequest('https://api.test.8base.com', 200, {
        data: {
          clientsList: {
            items: DATA.clients,
          },
        },
      }),
      global.mockRequest('https://api.test.8base.com', 200, {
        data: {
          ordersList: {
            items: DATA.orders,
          },
        },
      }),
    ];

    const workspaceData = await exportData(client.request.bind(client), TABLES);

    expect(await Promise.all(mocks)).toMatchSnapshot();
    expect(workspaceData).toMatchSnapshot();
  });

  it('should export data from workspace without system tables', async () => {
    const mocks = [
      global.mockRequest('https://api.test.8base.com', 200, tablesResponse),
      global.mockRequest('https://api.test.8base.com', 200, {
        data: {
          clientsList: {
            items: DATA.clients,
          },
        },
      }),
      global.mockRequest('https://api.test.8base.com', 200, {
        data: {
          ordersList: {
            items: DATA.orders,
          },
        },
      }),
      global.mockRequest('https://api.test.8base.com', 200, {
        data: {
          filesList: {
            items: [],
          },
        },
      }),
    ];

    const workspaceData = await exportData(client.request.bind(client), USER_TABLES);

    expect(await Promise.all(mocks)).toMatchSnapshot();
    expect(workspaceData).toMatchSnapshot();
  });

  describe('should have ability to choose which tables to export', () => {
    it('include', async () => {
      const mocks = [
        global.mockRequest('https://api.test.8base.com', 200, {
          data: {
            filesList: {
              items: [],
            },
          },
        }),
        global.mockRequest('https://api.test.8base.com', 200, {
          data: {
            clientsList: {
              items: DATA.clients,
            },
          },
        }),
      ];

      const workspaceData = await exportData(client.request.bind(client), USER_TABLES, {
        tables: {
          include: [CLIENTS_TABLE.name],
        },
      });

      expect(await Promise.all(mocks)).toMatchSnapshot();
      expect(workspaceData).toMatchSnapshot();
    });

    it('exclude', async () => {
      const mocks = [
        global.mockRequest('https://api.test.8base.com', 200, tablesResponse),
        global.mockRequest('https://api.test.8base.com', 200, {
          data: {
            ordersList: {
              items: DATA.orders,
            },
          },
        }),
        global.mockRequest('https://api.test.8base.com', 200, {
          data: {
            filesList: {
              items: [],
            },
          },
        }),
      ];

      const workspaceData = await exportData(client.request.bind(client), USER_TABLES, {
        tables: {
          exclude: [CLIENTS_TABLE.name],
        },
      });

      expect(await Promise.all(mocks)).toMatchSnapshot();
      expect(workspaceData).toMatchSnapshot();
    });
  });
});
