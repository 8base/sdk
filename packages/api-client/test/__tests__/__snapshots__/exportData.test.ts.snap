// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportData should export data from workspace with system tables 1`] = `
Array [
  Object {
    "body": Object {
      "query": "
      query UsersList {
        usersList {
          items {
            
            client {
              items {
                id
              }
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "244",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query FilesList {
        filesList {
          items {
            id
downloadUrl
public
filename
meta
mods

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "192",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query clientsList {
        clientsList {
          items {
            
            
            id
user {
              id
            }
fullName
orders {
              items {
                id
              }
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "317",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query ordersList {
        ordersList {
          items {
            
            id
client {
              id
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "208",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
]
`;

exports[`exportData should export data from workspace with system tables 2`] = `
Object {
  "Users": Array [
    Object {
      "$id": "user-1",
      "client": Object {
        "$id": "client-1",
      },
    },
  ],
  "clients": Array [
    Object {
      "$id": "client-1",
      "fullName": "Jacalyn Mccarrel",
      "orders": Array [
        Object {
          "$id": "order-1",
        },
        Object {
          "$id": "order-2",
        },
      ],
      "user": Object {
        "$id": "$currentUserId",
      },
    },
  ],
  "orders": Array [
    Object {
      "$id": "order-1",
      "client": Object {
        "$id": "client-1",
      },
    },
    Object {
      "$id": "order-2",
      "client": Object {
        "$id": "client-1",
      },
    },
  ],
}
`;

exports[`exportData should export data from workspace without system tables 1`] = `
Array [
  Object {
    "body": Object {
      "query": "
  query TableList($filter: TableListFilter) {
    tablesList(filter: $filter) {
      items {
        ...TableFragment
      }
    }
  }

  fragment TableFragment on Table {
    id
    name
    displayName
    isSystem
    fields {
      ...TableFieldFragment
    }
  }

  fragment TableFieldFragment on TableField {
    ...CommonTableFieldFragment
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
      ...SmartFieldTypesAttributes
    }
  }

  fragment CommonTableFieldFragment on TableField {
    id
    name
    displayName
    description
    fieldType
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
    }
    isList
    isRequired
    isUnique
    defaultValue
    isSystem
    isMeta
    relation {
      refFieldName
      refFieldDisplayName
      relationTableName
      relationFieldName
      refTable {
        id
        name
      }
      refFieldIsList
      refFieldIsRequired
    }
  }

  fragment DateFieldTypeAttributes on DateFieldTypeAttributes {
    format
  }

  fragment TextFieldTypeAttributes on TextFieldTypeAttributes {
    format
    fieldSize
  }

  fragment NumberFieldTypeAttributes on NumberFieldTypeAttributes {
    format
    precision
    currency
    minValue
    maxValue
    isBigInt
  }

  fragment FileFieldTypeAttributes on FileFieldTypeAttributes {
    format
    maxSize
    typeRestrictions
  }

  fragment SmartFieldTypesAttributes on SmartFieldTypeAttributes {
    format
    innerFields {
      name
      displayName
      description
      fieldType
      isList
      isRequired
      isUnique
      fieldTypeAttributes {
        ...TextFieldTypeAttributes
        ...NumberFieldTypeAttributes
        ...FileFieldTypeAttributes
        ...DateFieldTypeAttributes
        ...SwitchFieldTypeAttributes
      }
    }
  }

  fragment SwitchFieldTypeAttributes on SwitchFieldTypeAttributes {
    format
    listOptions
  }
",
      "variables": Object {
        "filter": Object {
          "tableNames": Array [
            "Files",
          ],
        },
      },
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "2338",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query clientsList {
        clientsList {
          items {
            
            
            id
user {
              id
            }
fullName
orders {
              items {
                id
              }
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "317",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query ordersList {
        ordersList {
          items {
            
            id
client {
              id
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "208",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query FilesList {
        filesList {
          items {
            id
downloadUrl
public
filename
meta
mods

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "192",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
]
`;

exports[`exportData should export data from workspace without system tables 2`] = `
Object {
  "Files": Array [],
  "clients": Array [
    Object {
      "$id": "client-1",
      "fullName": "Jacalyn Mccarrel",
      "orders": Array [
        Object {
          "$id": "order-1",
        },
        Object {
          "$id": "order-2",
        },
      ],
      "user": Object {
        "$id": "$currentUserId",
      },
    },
  ],
  "orders": Array [
    Object {
      "$id": "order-1",
      "client": Object {
        "$id": "client-1",
      },
    },
    Object {
      "$id": "order-2",
      "client": Object {
        "$id": "client-1",
      },
    },
  ],
}
`;

exports[`exportData should have ability to choose which tables to export exclude 1`] = `
Array [
  Object {
    "body": Object {
      "query": "
  query TableList($filter: TableListFilter) {
    tablesList(filter: $filter) {
      items {
        ...TableFragment
      }
    }
  }

  fragment TableFragment on Table {
    id
    name
    displayName
    isSystem
    fields {
      ...TableFieldFragment
    }
  }

  fragment TableFieldFragment on TableField {
    ...CommonTableFieldFragment
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
      ...SmartFieldTypesAttributes
    }
  }

  fragment CommonTableFieldFragment on TableField {
    id
    name
    displayName
    description
    fieldType
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
    }
    isList
    isRequired
    isUnique
    defaultValue
    isSystem
    isMeta
    relation {
      refFieldName
      refFieldDisplayName
      relationTableName
      relationFieldName
      refTable {
        id
        name
      }
      refFieldIsList
      refFieldIsRequired
    }
  }

  fragment DateFieldTypeAttributes on DateFieldTypeAttributes {
    format
  }

  fragment TextFieldTypeAttributes on TextFieldTypeAttributes {
    format
    fieldSize
  }

  fragment NumberFieldTypeAttributes on NumberFieldTypeAttributes {
    format
    precision
    currency
    minValue
    maxValue
    isBigInt
  }

  fragment FileFieldTypeAttributes on FileFieldTypeAttributes {
    format
    maxSize
    typeRestrictions
  }

  fragment SmartFieldTypesAttributes on SmartFieldTypeAttributes {
    format
    innerFields {
      name
      displayName
      description
      fieldType
      isList
      isRequired
      isUnique
      fieldTypeAttributes {
        ...TextFieldTypeAttributes
        ...NumberFieldTypeAttributes
        ...FileFieldTypeAttributes
        ...DateFieldTypeAttributes
        ...SwitchFieldTypeAttributes
      }
    }
  }

  fragment SwitchFieldTypeAttributes on SwitchFieldTypeAttributes {
    format
    listOptions
  }
",
      "variables": Object {
        "filter": Object {
          "tableNames": Array [
            "Files",
          ],
        },
      },
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "2338",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query ordersList {
        ordersList {
          items {
            
            id
client {
              id
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "208",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query FilesList {
        filesList {
          items {
            id
downloadUrl
public
filename
meta
mods

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "192",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
]
`;

exports[`exportData should have ability to choose which tables to export exclude 2`] = `
Object {
  "Files": Array [],
  "orders": Array [
    Object {
      "$id": "order-1",
      "client": Object {
        "$id": "client-1",
      },
    },
    Object {
      "$id": "order-2",
      "client": Object {
        "$id": "client-1",
      },
    },
  ],
}
`;

exports[`exportData should have ability to choose which tables to export include 1`] = `
Array [
  Object {
    "body": Object {
      "query": "
  query TableList($filter: TableListFilter) {
    tablesList(filter: $filter) {
      items {
        ...TableFragment
      }
    }
  }

  fragment TableFragment on Table {
    id
    name
    displayName
    isSystem
    fields {
      ...TableFieldFragment
    }
  }

  fragment TableFieldFragment on TableField {
    ...CommonTableFieldFragment
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
      ...SmartFieldTypesAttributes
    }
  }

  fragment CommonTableFieldFragment on TableField {
    id
    name
    displayName
    description
    fieldType
    fieldTypeAttributes {
      ...TextFieldTypeAttributes
      ...NumberFieldTypeAttributes
      ...FileFieldTypeAttributes
      ...DateFieldTypeAttributes
      ...SwitchFieldTypeAttributes
    }
    isList
    isRequired
    isUnique
    defaultValue
    isSystem
    isMeta
    relation {
      refFieldName
      refFieldDisplayName
      relationTableName
      relationFieldName
      refTable {
        id
        name
      }
      refFieldIsList
      refFieldIsRequired
    }
  }

  fragment DateFieldTypeAttributes on DateFieldTypeAttributes {
    format
  }

  fragment TextFieldTypeAttributes on TextFieldTypeAttributes {
    format
    fieldSize
  }

  fragment NumberFieldTypeAttributes on NumberFieldTypeAttributes {
    format
    precision
    currency
    minValue
    maxValue
    isBigInt
  }

  fragment FileFieldTypeAttributes on FileFieldTypeAttributes {
    format
    maxSize
    typeRestrictions
  }

  fragment SmartFieldTypesAttributes on SmartFieldTypeAttributes {
    format
    innerFields {
      name
      displayName
      description
      fieldType
      isList
      isRequired
      isUnique
      fieldTypeAttributes {
        ...TextFieldTypeAttributes
        ...NumberFieldTypeAttributes
        ...FileFieldTypeAttributes
        ...DateFieldTypeAttributes
        ...SwitchFieldTypeAttributes
      }
    }
  }

  fragment SwitchFieldTypeAttributes on SwitchFieldTypeAttributes {
    format
    listOptions
  }
",
      "variables": Object {
        "filter": Object {
          "tableNames": Array [
            "Files",
          ],
        },
      },
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "2338",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
  Object {
    "body": Object {
      "query": "
      query clientsList {
        clientsList {
          items {
            
            
            id
user {
              id
            }
fullName
orders {
              items {
                id
              }
            }

          }
        }
      }
    ",
      "variables": Object {},
    },
    "headers": Object {
      "accept": Array [
        "*/*",
      ],
      "accept-encoding": Array [
        "gzip,deflate",
      ],
      "authorization": Array [
        "Bearer idToken",
      ],
      "connection": Array [
        "close",
      ],
      "content-length": Array [
        "317",
      ],
      "content-type": Array [
        "application/json",
      ],
      "host": "api.test.8base.com",
      "user-agent": Array [
        "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      ],
      "workspace": Array [
        "workspaceId",
      ],
    },
  },
]
`;

exports[`exportData should have ability to choose which tables to export include 2`] = `
Object {
  "clients": Array [
    Object {
      "$id": "client-1",
      "fullName": "Jacalyn Mccarrel",
      "orders": Array [
        Object {
          "$id": "order-1",
        },
        Object {
          "$id": "order-2",
        },
      ],
      "user": Object {
        "$id": "$currentUserId",
      },
    },
  ],
}
`;
